components/core/src/clp/streaming_archive/single_file_archive/writer.cpp

```diff
@@ -0,0 +1,224 @@
+ * Combines file section metadata and the number of segments into single-file archive metadata,
+ * then serializes the metadata into MsgPack format.
+ *
+ * @param multi_file_archive_path
+ * @param num_segments
+ * @return Packed metadata.
+ */
+[[nodiscard]] auto pack_single_file_archive_metadata(
+        std::filesystem::path const& multi_file_archive_path,
+        size_t num_segments
+) -> std::stringstream;
+
+/**
+ * Writes single-file archive header.
+ *
+ * @param single_file_archive_writer
+ * @param packed_metadata_size
+ */
+auto write_archive_header(FileWriter& single_file_archive_writer, size_t packed_metadata_size)
+        -> void;
+
+/**
+ * Writes packed single-file archive metadata.
+ *
+ * @param single_file_archive_writer
+ * @param packed_metadata Packed metadata.
+ */
+auto write_packed_archive_metadata(
+        FileWriter& single_file_archive_writer,
+        std::stringstream const& packed_metadata
+) -> void;
+
+/**
+ * Reads the content of a file and writes it to the single-file archive.
+ * @param file_path
+ * @param single_file_archive_writer
+ * @throws OperationFailed if reading the file fails.
+ */
+auto
+write_archive_file(std::filesystem::path const& file_path, FileWriter& single_file_archive_writer)
+        -> void;
+
+/**
+ * Iterates over files in the multi-file archive and copies their contents to the single-file
+ * archive.
+ *
+ * @param single_file_archive_writer
+ * @param multi_file_archive_path
+ * @param num_segments
+ */
+auto write_archive_files(
+        FileWriter& single_file_archive_writer,
+        std::filesystem::path const& multi_file_archive_path,
+        size_t num_segments
+) -> void;
+
+auto
+get_archive_file_infos(std::filesystem::path const& multi_file_archive_path, size_t num_segments)
+        -> std::vector<FileInfo> {
+    std::vector<FileInfo> files;
+    uint64_t offset{};
+
+    for (auto const& static_archive_file_name : cStaticArchiveFileNames) {
+        files.emplace_back(FileInfo{std::string(static_archive_file_name), offset});
+        offset += std::filesystem::file_size(multi_file_archive_path / static_archive_file_name);
+    }
+
+    std::filesystem::path segment_dir_path = multi_file_archive_path / cSegmentsDirname;
+
+    for (size_t i = 0; i < num_segments; ++i) {
+        auto const segment_id = std::to_string(i);
+        files.emplace_back(FileInfo{segment_id, offset});
+        offset += std::filesystem::file_size(segment_dir_path / segment_id);
+    }
+
+    // Add sentinel indicating total size of all files.
+    files.emplace_back(FileInfo{std::string(cFileInfoSentinelName), offset});
+
+    return files;
+}
+
+auto pack_single_file_archive_metadata(
+        std::filesystem::path const& multi_file_archive_path,
+        size_t num_segments
+) -> std::stringstream {
+    SingleFileArchiveMetadata single_file_archive{
+            .archive_files = get_archive_file_infos(multi_file_archive_path, num_segments),
+            .num_segments = num_segments,
```

Comment:
Since `archive_files` and `num_segments` both use `num_segments`, they are inherently related; this means `SingleFileArchiveMetadata` should be a class.